crates/telio-traversal/src/endpoint_providers/upnp/mod.rs

```diff
@@ -132,43 +168,69 @@ impl UpnpEpCommands for IgdGateway {
             Some(gw) => gw,
             None => return Err(Error::NoIGDGateway),
         };
-        new_ext_port.0 = gw.add_any_port(
-            PortMappingProtocol::UDP,
-            SocketAddrV4::new(ip_addr, proxy_port),
-            3600,
-            "libminiupnp",
-        )?;
-        new_ext_port.1 = gw.add_any_port(
-            PortMappingProtocol::UDP,
-            SocketAddrV4::new(ip_addr, wg_port),
-            3600,
-            "libminiupnp",
-        )?;
+        new_ext_port.0 = gw
+            .add_any_port(
+                PortMappingProtocol::UDP,
+                SocketAddrV4::new(ip_addr, proxy_port),
+                3600,
+                "libminiupnp",
+            )
+            .await?;
+        new_ext_port.1 = gw
+            .add_any_port(
+                PortMappingProtocol::UDP,
+                SocketAddrV4::new(ip_addr, wg_port),
+                3600,
+                "libminiupnp",
+            )
+            .await?;
         Ok(new_ext_port)
     }
 
-    fn delete_endpoint_routes(&self, proxy_port: u16, wg_port: u16) -> Result<()> {
+    async fn delete_endpoint_routes(&self, proxy_port: u16, wg_port: u16) -> Result<()> {
         let gw = match &self.gw {
             Some(gw) => gw,
             None => return Err(Error::NoIGDGateway),
         };
-        gw.remove_port(PortMappingProtocol::UDP, wg_port)?;
-        gw.remove_port(PortMappingProtocol::UDP, proxy_port)?;
+        gw.remove_port(PortMappingProtocol::UDP, wg_port).await?;
+        gw.remove_port(PortMappingProtocol::UDP, proxy_port).await?;
         Ok(())
     }
 
-    fn get_igd_gateway(&mut self) -> Result<()> {
-        self.gw = Some(search_gateway(Default::default())?);
-        Ok(())
+    async fn get_igd_gateway(&mut self) -> Result<()> {
+        if let Some(handle) = &mut self.search_handle {
+            tokio::select! {
+                _ = sleep(Duration::from_secs(GET_IGW_GATEWAY_TIMEOUT_MS)) => {},
```

Comment:
I think you have a bug here, at least judging by the names (secs vs. MS). That's precisely why we should avoid keeping time as u64 :D

Reply:
Oh you're right, missed that